pgrx/src/pg_catalog.rs

```diff
@@ -0,0 +1,866 @@
+        unsafe { CStr::from_ptr(raw.cast::<c_char>()) }
+    }
+}
+
+unsafe impl GetStruct<pgrx::pg_sys::int2vector> for &[i16] {
+    unsafe fn get_struct(raw: *const pgrx::pg_sys::int2vector) -> Self {
+        unsafe { (*raw).values.as_slice((*raw).dim1 as usize) }
+    }
+}
+
+unsafe impl GetStruct<pgrx::pg_sys::oidvector> for &[Oid] {
+    unsafe fn get_struct(raw: *const pgrx::pg_sys::oidvector) -> Self {
+        unsafe { (*raw).values.as_slice((*raw).dim1 as usize) }
+    }
+}
+
+macro_rules! _macro_1 {
+    {
+        $table:ident, ($column:ident, character { $(($variant:ident, $value:literal))* } $($x:tt)*)
+    } => {
+        paste::paste! {
+            #[non_exhaustive]
+            #[repr(u8)]
+            #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+            pub enum [<$table:camel $column:camel>] {
+                $($variant = $value),*
+            }
+
+            impl [<$table:camel $column:camel>] {
+                fn from_c_char(value: c_char) -> Self {
+                    match value as u8 {
+                        $($value => Self::$variant,)*
+
+                        _ => panic!("unrecognized value: `{}`", value as u8 as char),
+                    }
+                }
+            }
+
+            impl pgrx::datum::FromDatum for [<$table:camel $column:camel>] {
+                unsafe fn from_polymorphic_datum(
+                    datum: pgrx::pg_sys::Datum,
+                    is_null: bool,
+                    _: Oid,
+                ) -> Option<Self>
+                where
+                    Self: Sized,
+                {
+                    if is_null {
+                        None
+                    } else {
+                        Some(Self::from_c_char(datum.value() as _))
+                    }
+                }
+            }
+
+            impl pgrx::datum::IntoDatum for [<$table:camel $column:camel>] {
+                fn into_datum(self) -> std::option::Option<pgrx::pg_sys::Datum> {
+                    Some(pgrx::pg_sys::Datum::from(self as i8))
+                }
+                fn type_oid() -> pgrx_pg_sys::Oid {
+                    pgrx::pg_sys::CHAROID
+                }
+            }
+
+            unsafe impl pgrx::datum::UnboxDatum for [<$table:camel $column:camel>] {
+                type As<'src> = Self;
+                #[inline]
+                unsafe fn unbox<'src>(datum: pgrx::datum::Datum<'src>) -> Self::As<'src>
+                where
+                    Self: 'src,
+                {
+                    Self::from_c_char(datum.sans_lifetime().value() as c_char)
+                }
+            }
+
+            unsafe impl GetStruct<c_char> for [<$table:camel $column:camel>] {
+                unsafe fn get_struct(raw: *const c_char) -> Self {
+                    unsafe { Self::from_c_char(raw.read()) }
+                }
+            }
+        }
+    };
+}
+
+macro_rules! _macro_0 {
```

Comment:
These macros definitely need names, not ordinals.

Reply:
Renamed to `define_catalog_column`.